src/smt/proof_post_processor.cpp

```diff
@@ -130,6 +130,245 @@ bool ProofPostprocessCallback::updateInternal(Node res,
+      // position of the last resolution link that introduces the crowding
+      // literal. Note that this position has to be *before* the last link, as a
+      // link *after* the last inclusion must eliminate the crowding literal.
+      size_t j;
+      for (j = children.size() - 1; j > 0; --j)
+      {
+        // notice that only non-unit clauses may be introducing the crowding
+        // literal, so we don't need to differentiate unit from non-unit
+        if (children[j - 1].getKind() != kind::OR)
+        {
+          continue;
+        }
+        if (std::find(children[j - 1].begin(), children[j - 1].end(), crowdLit)
+            != children[j - 1].end())
+        {
+          break;
+        }
+      }
+      Assert(j > 0);
+      lastInclusion.push_back(std::make_pair(crowdLit, j - 1));
+      Trace("smt-proof-pp-debug2") << "crowding lit " << crowdLit << "\n";
+      Trace("smt-proof-pp-debug2") << "last inc " << j - 1 << "\n";
+      // get elimination position, starting from the following link as the last
+      // inclusion one. The reslut is the last resolution link that eliminates
+      // the crowding literal. A literal l is eliminated by a link if it
+      // contains a literal l' with opposite polarity to l.
+      for (; j < children.size(); ++j)
+      {
+        bool posFirst = args[(2 * j) - 1] == trueNode;
+        Node pivot = args[(2 * j)];
+        Trace("smt-proof-pp-debug2")
+            << "\tcheck w/ args " << posFirst << " / " << pivot << "\n";
+        // To eliminate the crowding literal (crowdLit), the clause must contain
+        // it with opposity polarity. There are three successful cases,
+        // according to the pivot and its sign
+        //
+        // - crowdLit is the same as the pivot and posFirst is true, which means
+        //   that the clause contains its negation and eliminates it
+        //
+        // - the pivot is equal to crowdLit negated and posFirst is false, which
+        //   means that the clause contains the negation of crowdLit
+        //
+        // - crowdLit is the negation of the pivot and posFirst is false, so the
+        //   clause contains the node whose negation is crowdLit
+        if ((crowdLit == pivot && posFirst)
+            || (crowdLit.notNode() == pivot && !posFirst)
+            || (pivot.notNode() == crowdLit && !posFirst))
+        {
+          Trace("smt-proof-pp-debug2") << "\t\tfound it!\n";
+          eliminators.push_back(j);
+          break;
+        }
+      }
+      Assert(j < children.size());
+    }
+  }
+  Assert(!lastInclusion.empty());
+  // order map so that we process crowding literals in the order of the clauses
+  // that last introduce them
+  auto cmp = [=](std::pair<Node, size_t>& a, std::pair<Node, size_t>& b) {
+    return a.second < b.second;
+  };
+  std::sort(lastInclusion.begin(), lastInclusion.end(), cmp);
+  // order eliminators
+  std::sort(eliminators.begin(), eliminators.end());
```

Comment:
I'm a bit confused about the fact that we can simply sort this. We initially compute `lastInclusion` and `eliminators` both for a specific literal, i.e. there is a certain correspondence for the entries in these two vectors. Sorting them independently loses this correspondence... is this safe?

Reply:
The sorting happens because we build these two things according to how literals appear in `clauseLits`, which does not necessarily correspond to the order in which the resolutions happen. The sorting is safe because both `lastInclusion` and `eliminators` are being sorted according to the positions in the premises, and the invariant when they're being built is that the last inclusions occur in between the eliminators. Sorting actually gives us the organized view.